### ISSUE TITLE
False negative with vue/require-valid-default-prop on props with type assertion

### ISSUE THREAD

**Issue by frozenbonito** [vuejs/eslint-plugin-vue#1468]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 7.23.0
- **eslint-plugin-vue version:** 7.8.0 
- **Node version:** 14.16.0
- **Operating System:** Windows 10 Pro

**Please show your full configuration:**

```js
module.exports = {
  root: true,
  env: {
    node: true,
  },
  extends: [
    "plugin:vue/essential",
    "eslint:recommended",
    "@vue/typescript/recommended",
    "@vue/prettier",
    "@vue/prettier/@typescript-eslint",
  ],
  parserOptions: {
    ecmaVersion: 2020,
  },
  rules: {},
};
```

**What did you do?**	

I annotated type to array prop with typescript according to the following documentations.
- https://vuejs.org/v2/guide/typescript.html#Annotating-Props
- https://v3.vuejs.org/guide/typescript-support.html#annotating-return-types

If the prop is an array, the default should be a function, but I set empty array directly by mistake.

```vue
<script lang="ts">
import Vue from "vue";
import { PropType } from "vue";

export default Vue.extend({
  props: {
    foo: {
      type: Array as PropType<string[]>,
      default: [],
    },
  },
});
</script>
```

**What did you expect to happen?**

eslint reports type of the default value for 'foo' prop must be a function.

**What actually happened?**

eslint did not report that. If there is no type assertion, it reports correctly.

**Comment by VRuzhentsov** [comment]
Will this work with the [next case](http://dl3.joxi.net/drive/2021/04/05/0030/1408/1971584/84/92890a21bf.png)?
